storage/aptosdb/src/pruner/worker.rs

```diff
@@ -0,0 +1,140 @@
+// Copyright (c) The Diem Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+use aptos_types::transaction::Version;
+use schemadb::DB;
+
+use crate::pruner::{db_pruner::DBPruner, transaction_store::TransactionStorePruner};
+use aptos_infallible::Mutex;
+
+use crate::{pruner::state_store::StateStorePruner, TransactionStore};
+use itertools::zip_eq;
+use std::{
+    sync::{mpsc::Receiver, Arc},
+    time::Instant,
+};
+
+/// Maintains all the DBPruners and periodically calls the db_pruner's prune method to prune the DB.
+/// This also exposes API to report the progress to the parent thread.
+pub struct Worker {
+    command_receiver: Receiver<Command>,
+    /// Keeps tracks of all the DB pruners
+    db_pruners: Vec<Mutex<Arc<dyn DBPruner + Send + Sync>>>,
+    /// Keeps a record of the pruning progress. If this equals to version `V`, we know versions
+    /// smaller than `V` are no longer readable.
+    /// This being an atomic value is to communicate the info with the Pruner thread (for tests).
+    least_readable_versions: Arc<Mutex<Vec<Version>>>,
+    /// Indicates if there's NOT any pending work to do currently, to hint
+    /// `Self::receive_commands()` to `recv()` blocking-ly.
+    blocking_recv: bool,
+}
+
+impl Worker {
+    const MAX_VERSIONS_TO_PRUNE_PER_BATCH: usize = 100;
+
+    pub(crate) fn new(
+        db: Arc<DB>,
+        transaction_store: Arc<TransactionStore>,
+        command_receiver: Receiver<Command>,
+        least_readable_versions: Arc<Mutex<Vec<Version>>>,
+    ) -> Self {
+        Self {
+            db_pruners: vec![
+                Mutex::new(Arc::new(StateStorePruner::new(
+                    Arc::clone(&db),
+                    0,
+                    Instant::now(),
+                ))),
+                Mutex::new(Arc::new(TransactionStorePruner::new(
+                    Arc::clone(&db),
+                    Arc::clone(&transaction_store),
+                ))),
+            ],
+            command_receiver,
+            least_readable_versions,
+            blocking_recv: true,
+        }
+    }
+
+    pub(crate) fn work(mut self) {
+        for db_pruner in &self.db_pruners {
+            db_pruner.lock().initialize();
+        }
+        while self.receive_commands() {
+            // Process a reasonably small batch of work before trying to receive commands again,
+            // in case `Command::Quit` is received (that's when we should quit.)
+            for db_pruner in &self.db_pruners {
+                let result = db_pruner
+                    .lock()
+                    .prune(Self::MAX_VERSIONS_TO_PRUNE_PER_BATCH);
+                match result {
+                    Ok(_) => {}
+                    Err(_) => {
+                        // On error, stop retrying vigorously by making next recv() blocking.
+                        self.blocking_recv = true;
```

Comment:
We still need the logic of "if all pruners has nothing to do, set recv_blocking to true"? Did I miss it?

Reply:
Yes you are right. Actually this logic is a bit tricky with multiple pruners involved. For example, what should we do if there is pruning pending from one pruner but error from another. I have simplified the logic to ignore errors for now. LMK what you think.

Reply:
Yeah the error is not likely, I think the current version looks good.